xdsl/transforms/mlir_opt.py

```diff
@@ -0,0 +1,51 @@
+import shutil
+import subprocess
+from dataclasses import dataclass, field
+from io import StringIO
+
+from xdsl.dialects.builtin import ModuleOp
+from xdsl.ir.core import MLContext
+from xdsl.parser.core import Parser
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import PatternRewriter
+from xdsl.printer import Printer
+
+
+@dataclass
+class MLIROptPass(ModulePass):
+    """
+    A pass for calling the `mlir-opt` tool with specified parameters. Will fail if
+    `mlir-opt` is not available.
+    """
+
+    name = "mlir-opt"
+
+    arguments: list[str] = field(default_factory=list)
+
+    def apply(self, ctx: MLContext, op: ModuleOp) -> None:
+        if not shutil.which("mlir-opt"):
+            raise ValueError("mlir-opt is not available")
+
+        stream = StringIO()
+        printer = Printer(print_generic_format=True, stream=stream)
+        printer.print(op)
+
+        my_string = stream.getvalue()
+
+        completed_process = subprocess.run(
+            ["mlir-opt", *self.arguments],
+            input=my_string,
+            stdout=subprocess.PIPE,
+            text=True,
+        )
+
+        # Get the stdout output
+        stdout_output = completed_process.stdout
```

Comment:
Can you look at `completed_process.returncode`, and return an exception if it is not 0?
That way, if mlir-opt fails, we will see it.

Reply:
- [ ] return code